### ISSUE TITLE
Vision encoder weights does not seem to be updated

### ISSUE THREAD

**Issue by fedshyvana** [2U1/Qwen2-VL-Finetune#115]
Thank you for contributing this wonderful code base!

I ran two experiments with the following understanding:

Exp 1: Update LLM with Lora, keep vision encoder frozen + finetune merger 
```
    --use_liger True \
    --lora_enable True \
    --use_dora False \
    --lora_namespan_exclude "['lm_head', 'embed_tokens']" \
    --lora_rank 64 \
    --lora_alpha 64 \
    --lora_dropout 0.05 \
    --num_lora_modules -1 \
    --freeze_vision_tower True \
    --freeze_llm True \
    --freeze_merger False \
```

Exp 2: Update LLM with Lora, finetune vision encoder + merger
```
    --use_liger True \
    --lora_enable True \
    --use_dora False \
    --lora_namespan_exclude "['lm_head', 'embed_tokens']" \
    --lora_rank 64 \
    --lora_alpha 64 \
    --lora_dropout 0.05 \
    --num_lora_modules -1 \
    --freeze_vision_tower False \
    --freeze_llm True \
    --freeze_merger False \
```

However, for the same checkpoint (e.g. iterations 250), the loss values / VRAM usage for the two runs are nearly identical, furthermore, when I inspect the weights across the two runs in `non_lora_state_dict.bin` using torch.allclose, it returns True for everything. This suggests to me either the vision encoder is not frozen correctly in the 1st run or it was not unfrozen correctly in the 2nd run. Could you please let me know if my understanding is correct? Thank you!

**Comment by fedshyvana** [comment]
Actually i just found that there's also no change in weights in `non_lora_state_dict.bin` between iterations 250 and 500 of the first run (i.e. only merger is tuned). Where is the updated weights being saved then?

**Comment by 2U1** [comment]
Thanks for letting me know. The vision encoder weights should be saved in `non_lora_state.bin`. 
I'll take a look in the code.

**Comment by fedshyvana** [comment]
Thanks @2U1! Yeah I can confirm the weights in `non_lora_state.bin` remain the same across e.g. 250 vs. 500 vs. 1000 iterations in both Exp 1 and Exp 2 (i.e. whether i have `--freeze_merger False` or both `--freeze_merger False` and `--freeze_vision_tower False`). I am using:
- transformers: 4.50.1
- accelerate: 1.3.0
- torch: 2.5.1

Compared to your provided conda env file, the only difference is that I first ran `pip install torch==2.5.1 torchvision==0.20.1 torchaudio==2.5.1 --index-url https://download.pytorch.org/whl/cu118` since I have CUDA 11.8 instead of 12 (using the environment.yaml directly leads to deepspeed complaining that my torch is compiled with CUDA 12).

**Comment by fedshyvana** [comment]
@2U1 I still need to verify this but I believe the issue might be because Peft LoRA model will set all non-lora parameters to requires_grad to be False. As a result, even though the vision tower / merger weights have been configured to be unfrozen earlier in the training script, they are once again frozen when get_peft_model gets called.

**Comment by 2U1** [comment]
@fedshyvana I've checked the code and you were right. 
At the first version of the code, the vision_tower was unforzen despite I've set the freeze and using lora, but it has the opposite problem now.

I've fixed the code and ;it should be properly unfreezed/freezed.
